In [66]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px


In [8]:
data = pd.read_csv('open_meteo.csv')
data

Unnamed: 0        time  weathercode  temperature_2m_max  \
0               0  2010-01-01            0                26.0   
1               1  2010-01-02            2                25.3   
2               2  2010-01-03            0                25.0   
3               3  2010-01-04            0                25.6   
4               4  2010-01-05            0                25.4   
...           ...         ...          ...                 ...   
34757       34757  2023-08-02           51                30.2   
34758       34758  2023-08-03           51                30.4   
34759       34759  2023-08-04           51                30.5   
34760       34760  2023-08-05           51                30.7   
34761       34761  2023-08-06           51                30.7   

       temperature_2m_min  temperature_2m_mean  apparent_temperature_max  \
0                    14.8                 20.9                      24.0   
1                    15.6                 20.5                      21.8   
2                    14.3                 19.9                      22.8   
3                    15.0                 20.4                      23.9   
4                    16.3                 20.7                      24.5   
...                   ...                  ...                       ...   
34757                25.5                 27.4                      31.5   
34758                25.4                 27.3                      31.5   
34759                25.5                 27.5                      32.8   
34760                25.6                 27.5                      32.5   
34761                25.4                 27.3                      31.7   

       apparent_temperature_min  apparent_temperature_mean           sunrise  \
0                          13.4                       18.6  2010-01-01T02:16   
1                          13.7                       18.1  2010-01-02T02:17   
2                          14.2                       18.9  2010-01-03T02:17   
3                          13.7                       19.4  2010-01-04T02:17   
4                          16.8                       20.7  2010-01-05T02:17   
...                         ...                        ...               ...   
34757                      27.8                       29.3  2023-08-02T00:52   
34758                      28.2                       29.7  2023-08-03T00:52   
34759                      29.3                       30.6  2023-08-04T00:52   
34760                      28.9                       30.4  2023-08-05T00:53   
34761                      27.0                       28.9  2023-08-06T00:53   

       ... precipitation_hours  windspeed_10m_max  windgusts_10m_max  \
0      ...                 0.0               18.3               25.6   
1      ...                 0.0               24.6               35.3   
2      ...                 0.0               16.6               28.1   
3      ...                 0.0               14.0               22.7   
4      ...                 0.0               19.3               29.2   
...    ...                 ...                ...                ...   
34757  ...                 6.0               37.7               58.3   
34758  ...                10.0               36.4               56.2   
34759  ...                 6.0               31.8               49.3   
34760  ...                 6.0               30.7               48.6   
34761  ...                 4.0               38.8               60.5   

       winddirection_10m_dominant  et0_fao_evapotranspiration   latitude  \
0                              54                        4.38  24.900002   
1                              64                        3.84  24.900002   
2                             312                        3.69  24.900002   
3                             353                        3.63  24.900002   
4                             270                        3.55  24.900002   
...                

In [13]:
data.describe()

latitude  temperature_2m_min  temperature_2m_max  \
count  34762.000000        34762.000000        34762.000000   
mean      17.900002           23.205748           30.677150   
std        5.478609            3.615309            3.236343   
min        8.500000            6.800000           17.700000   
25%       12.900002           21.700000           28.300000   
50%       19.099998           24.200000           30.500000   
75%       23.300003           25.500000           32.800000   
max       24.900002           32.500000           44.300000   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
count         34762.000000       34762.000000         34762.000000   
mean             26.674475           4.496036             5.321788   
std               2.733634          12.391169             8.286665   
min              13.200000           0.000000             0.000000   
25%              25.500000           0.000000             0.000000   
50%              26.800000           0.000000             0.000000   
75%              28.300000           2.800000             9.000000   
max              37.100000         307.900000            24.000000   

       et0_fao_evapotranspiration  
count                34762.000000  
mean                     4.568428  
std                      1.553584  
min                      0.700000  
25%                      3.620000  
50%                      4.520000  
75%                      5.460000  
max                     13.280000

In [28]:
data = data[['time', 'city', 'temperature_2m_min', 'temperature_2m_max', 'temperature_2m_mean', 'precipitation_sum', 'precipitation_hours', 'et0_fao_evapotranspiration']]
data

time     city  temperature_2m_min  temperature_2m_max  \
0     2010-01-01  Karachi                14.8                26.0   
1     2010-01-02  Karachi                15.6                25.3   
2     2010-01-03  Karachi                14.3                25.0   
3     2010-01-04  Karachi                15.0                25.6   
4     2010-01-05  Karachi                16.3                25.4   
...          ...      ...                 ...                 ...   
34757 2023-08-02     Bhuj                25.5                30.2   
34758 2023-08-03     Bhuj                25.4                30.4   
34759 2023-08-04     Bhuj                25.5                30.5   
34760 2023-08-05     Bhuj                25.6                30.7   
34761 2023-08-06     Bhuj                25.4                30.7   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
0                     20.9                0.0                  0.0   
1                     20.5                0.0                  0.0   
2                     19.9                0.0                  0.0   
3                     20.4                0.0                  0.0   
4                     20.7                0.0                  0.0   
...                    ...                ...                  ...   
34757                 27.4                0.6                  6.0   
34758                 27.3                1.0                 10.0   
34759                 27.5                0.8                  6.0   
34760                 27.5                0.6                  6.0   
34761                 27.3                0.4                  4.0   

       et0_fao_evapotranspiration  
0                            4.38  
1                            3.84  
2                            3.69  
3                            3.63  
4                            3.55  
...                           ...  
34757                        4.07  
34758                        3.92  
34759                        3.75  
34760                        4.10  
34761                        4.31  

[34762 rows x 8 columns]

In [29]:
data['time'] = pd.to_datetime(data['time'], format='%Y/%m/%d')


C:\Users\ASUS\anaconda3\lib\site-packages\pandas\core\frame.py:3607: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._set_item(key, value)


In [33]:
start_date = '5-1-2022'
end_date = '12-1-2022'
df = data[(data['time'] >= start_date) & (data['time'] <= end_date)]

In [34]:
df

time     city  temperature_2m_min  temperature_2m_max  \
4503  2022-05-01  Karachi                28.0                36.2   
4504  2022-05-02  Karachi                26.6                33.1   
4505  2022-05-03  Karachi                26.1                33.5   
4506  2022-05-04  Karachi                25.5                32.9   
4507  2022-05-05  Karachi                25.0                33.0   
...          ...      ...                 ...                 ...   
34509 2022-11-27     Bhuj                15.4                29.8   
34510 2022-11-28     Bhuj                14.7                29.8   
34511 2022-11-29     Bhuj                14.1                29.4   
34512 2022-11-30     Bhuj                15.0                30.0   
34513 2022-12-01     Bhuj                14.9                29.0   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
4503                  31.8                0.0                  0.0   
4504                  29.5                0.0                  0.0   
4505                  29.4                0.0                  0.0   
4506                  28.7                0.0                  0.0   
4507                  28.3                0.0                  0.0   
...                    ...                ...                  ...   
34509                 22.8                0.0                  0.0   
34510                 22.3                0.0                  0.0   
34511                 21.8                0.0                  0.0   
34512                 22.6                0.0                  0.0   
34513                 22.2                0.0                  0.0   

       et0_fao_evapotranspiration  
4503                         9.70  
4504                         7.83  
4505                         6.71  
4506                         6.50  
4507                         6.32  
...                           ...  
34509                        4.05  
34510                        4.48  
34511                        5.08  
34512                        5.41  
34513                        4.81  

[1505 rows x 8 columns]

### Which city has low temperature and high temperature? 

In [52]:
low_temp = df['temperature_2m_min'].min()
df[df['temperature_2m_min'] == low_temp]

time  city  temperature_2m_min  temperature_2m_max  \
34511 2022-11-29  Bhuj                14.1                29.4   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
34511                 21.8                0.0                  0.0   

       et0_fao_evapotranspiration  
34511                        5.08

We can see that Bhuj has low temperature than the other cities.
The city was not too hot at 2022 November

In [54]:
high_temp = df['temperature_2m_max'].max()
df[df['temperature_2m_max'] == high_temp] 

time  city  temperature_2m_min  temperature_2m_max  \
34310 2022-05-12  Bhuj                25.2                42.2   
34353 2022-06-24  Bhuj                27.8                42.2   

       temperature_2m_mean  precipitation_sum  precipitation_hours  \
34310                 32.9                0.0                  0.0   
34353                 34.6                0.0                  0.0   

       et0_fao_evapotranspiration  
34310                       10.16  
34353                        8.29

From the above, The maximum temperature recorded in Bhuj was 42.2°C. The months of May and June are indeed known for higher temperatures and heatwaves in Bhuj.

### Minimum Temperature and Maximum Temperature Plot

In [78]:
fig = px.line(df, x="time", y="temperature_2m_min", title='Minimum Temperature for each day',color='city')
fig2 = px.line(df, x="time", y="temperature_2m_max", title='Maximum Temperature for each day',color='city')
fig3 = px.line(df, x="time", y="temperature_2m_mean", title='Average Temperature for each day',color='city')

fig.show()
fig2.show()
fig3.show()